// simba-kernel/lib/jshint.js
/**
 * Created by 14060208 on 2015/12/4.
 */
"use strict";

var fs = require('fs'),
    path = require('path'),
    simba = require('simba-kernel');

module.exports = {
    reporter: function(results, data, opts) {
        var len = results.length,
            str = '',
            prevfile,
            errorCount = 0,
            warningCount = 0,
            message = [],
            file = '',
            error = '',
            flag = false;
        opts = opts || {};

        results.forEach(function(result) {
            if (result) {
                if (/^E.*?/.test(result.code) && result.code != 'E043') {
                    result.type = 'error';
                    errorCount += 1;
                } else {
                    result.type = 'warning';
                    warningCount += 1;
                }
            }
            file = result.file;
            error = result.error;
            if (prevfile && prevfile !== file) {
                str += "\n";
            }
            prevfile = file;

            if(!flag){
                flag =true;
                console.log('\n'+file.blue.bold);
            }

            if(result.type == "error"){
                str = result.type + ': line ' + error.line + ', col ' +
                error.character + ', ' + error.reason;
                //console.log(str.red.bold);
            }else{
                str = result.type + ': line ' + error.line + ', col ' +
                error.character + ', ' + error.reason;
                //console.log(str.yellow.bold);
            }
        });

        message.push({
            messages: results,
            file: file
        });
        var root = path.join(path.resolve('./'), '/'),
            buildfile = root + '/report/report-jshint-'+file.substr(file.lastIndexOf("\\")+1,file.length)+'.html',  // 生成报告文件
            tpl = __dirname + '/report-template-jshint.html'; // 报告模板
        var reportDir = path.join(root, './report');
        if(fs.existsSync(reportDir+'/report-jshint-index.html')){
            simba.util.empty(root+'/report')
        }

        /*
        * 如果不存在，则建立 .report 文件夹
        * */
        if (!simba.util.exists(reportDir)) {
            fs.mkdirSync(reportDir);
        }

        var template = simba.template;
        template.config('extname', '');

        var d = simba.util.localDate();
        var html = template(tpl, {
            data: message,
            version: simba.info.version,
            time: d
        });
        // 写入报告文件
        fs.writeFileSync(buildfile, html);
        // 控制台提示报告已写入
        var file = buildfile.replace(root, '');
        console.log('\n Report file has been created: '.yellow.bold + (root+path.normalize(file).substr(1)).cyan.bold);

    }
};
